File: scripts/inject_context.py
#!/usr/bin/env python
"""
Elephants Never Forget — SessionStart Context Injector

Reads index.md, decisions.md, and the most recent session files from
.claude-sessions/ and outputs them as context for Claude.

Output goes to stdout and is injected into Claude's context window.

Token budget: aims to inject ~3,000-5,000 tokens total.
- index.md: up to 4,000 chars (~1,000 tokens)
- decisions.md: up to 4,000 chars (~1,000 tokens)
- Recent session files: up to 8,000 chars (~2,000 tokens) for the last 3
"""

import json
import sys
import os
import glob


def read_file_safe(path, max_chars=4000):
    """Read a file safely, truncating at a line boundary near max_chars."""
    try:
        with open(path, "r", encoding="utf-8") as f:
            content = f.read(max_chars + 200)  # Read a bit extra
        if len(content) > max_chars:
            # Truncate at last newline before max_chars
            truncated = content[:max_chars]
            last_nl = truncated.rfind("\n")
            if last_nl > 0:
                truncated = truncated[:last_nl]
            return truncated + "\n... (truncated — read full file for more)"
        return content
    except (FileNotFoundError, PermissionError):
        return ""


def get_recent_sessions(sessions_dir, max_files=3, max_chars_each=2700):
    """Read the most recent session files (by filename date sort)."""
    pattern = os.path.join(sessions_dir, "sessions", "*.md")
    files = sorted(glob.glob(pattern), reverse=True)  # Newest first by name

    results = []
    for fpath in files[:max_files]:
        content = read_file_safe(fpath, max_chars_each)
        if content:
            fname = os.path.basename(fpath)
            results.append(f"### {fname}\n{content}")
    return results


def main():
    try:
        input_data = json.load(sys.stdin)
        project_dir = os.environ.get("CLAUDE_PROJECT_DIR", os.getcwd())
        sessions_dir = os.path.join(project_dir, ".claude-sessions")
        session_id = input_data.get("session_id", "unknown")

        index_content = read_file_safe(os.path.join(sessions_dir, "index.md"))
        decisions_content = read_file_safe(os.path.join(sessions_dir, "decisions.md"))
        recent_sessions = get_recent_sessions(sessions_dir)

        if not index_content and not decisions_content and not recent_sessions:
            sys.exit(0)

        parts = []
        parts.append("=== ELEPHANTS NEVER FORGET: Cross-Session Context ===")
        parts.append(f"Current session: {session_id[:8]}")
        parts.append("")

        if index_content:
            parts.append("## Session Index")
            parts.append(index_content)
            parts.append("")

        if decisions_content:
            parts.append("## Standing Decisions")
            parts.append(decisions_content)
            parts.append("")

        if recent_sessions:
            parts.append("## Recent Sessions (Hot Tier — last 3)")
            parts.extend(recent_sessions)
            parts.append("")

        parts.append("---")
        parts.append("Use the elephants-never-forget skill to maintain this session's log.")

        print("\n".join(parts))

    except Exception:
        pass

    sys.exit(0)


if __name__ == "__main__":
    main()

File: scripts/inject_context.js
/**
 * Elephants Never Forget — SessionStart Context Injector
 *
 * Reads index.md, decisions.md, and the most recent session files from
 * .claude-sessions/ and outputs them as context for Claude.
 *
 * Output goes to stdout and is injected into Claude's context window.
 */
import { readFileSync, readdirSync } from 'node:fs';
import { join, basename } from 'node:path';

function readFileSafe(path, maxChars = 4000) {
  try {
    const full = readFileSync(path, 'utf8');
    const head = full.slice(0, maxChars + 200);
    if (head.length > maxChars) {
      let truncated = head.slice(0, maxChars);
      const lastNl = truncated.lastIndexOf('\n');
      if (lastNl > 0) truncated = truncated.slice(0, lastNl);
      return truncated + '\n... (truncated — read full file for more)';
    }
    return head;
  } catch {
    return '';
  }
}

function getRecentSessions(sessionsDir, maxFiles = 3, maxCharsEach = 2700) {
  const dir = join(sessionsDir, 'sessions');
  let files;
  try {
    files = readdirSync(dir).filter((f) => f.endsWith('.md')).sort().reverse();
  } catch {
    return [];
  }
  const results = [];
  for (const fname of files.slice(0, maxFiles)) {
    const content = readFileSafe(join(dir, fname), maxCharsEach);
    if (content) {
      results.push(`### ${fname}\n${content}`);
    }
  }
  return results;
}

function readStdin() {
  return new Promise((resolveStdin) => {
    let data = '';
    process.stdin.setEncoding('utf8');
    process.stdin.on('data', (chunk) => { data += chunk; });
    process.stdin.on('end', () => resolveStdin(data));
    process.stdin.on('error', () => resolveStdin(''));
  });
}

async function main() {
  try {
    const raw = await readStdin();
    const input = raw ? JSON.parse(raw) : {};
    const projectDir = process.env.CLAUDE_PROJECT_DIR || process.cwd();
    const sessionsDir = join(projectDir, '.claude-sessions');
    const sessionId = input.session_id || 'unknown';

    const indexContent = readFileSafe(join(sessionsDir, 'index.md'));
    const decisionsContent = readFileSafe(join(sessionsDir, 'decisions.md'));
    const recentSessions = getRecentSessions(sessionsDir);

    if (!indexContent && !decisionsContent && recentSessions.length === 0) {
      process.exit(0);
      return;
    }

    const parts = [];
    parts.push('=== ELEPHANTS NEVER FORGET: Cross-Session Context ===');
    parts.push(`Current session: ${String(sessionId).slice(0, 8)}`);
    parts.push('');

    if (indexContent) {
      parts.push('## Session Index');
      parts.push(indexContent);
      parts.push('');
    }

    if (decisionsContent) {
      parts.push('## Standing Decisions');
      parts.push(decisionsContent);
      parts.push('');
    }

    if (recentSessions.length > 0) {
      parts.push('## Recent Sessions (Hot Tier — last 3)');
      for (const s of recentSessions) parts.push(s);
      parts.push('');
    }

    parts.push('---');
    parts.push('Use the elephants-never-forget skill to maintain this session\'s log.');

    process.stdout.write(parts.join('\n') + '\n');
  } catch {
    // swallow
  }
  process.exit(0);
}

main();
